pkg/kv/postgres/store.go

```diff
@@ -0,0 +1,239 @@
+	}
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", kv.ErrOperationFailed, err)
+	}
+	return val, nil
+}
+
+func (s *Store) Set(ctx context.Context, key, value []byte) error {
+	if key == nil {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+	_, err := s.Pool.Exec(ctx, `INSERT INTO `+s.Params.TableName+`(key,value) VALUES($1,$2)
+			ON CONFLICT (key) DO UPDATE SET value = $2`, key, value)
+	if err != nil {
+		return fmt.Errorf("%w: %s", kv.ErrOperationFailed, err)
+	}
+	return nil
+}
+
+func (s *Store) SetIf(ctx context.Context, key, value, valuePredicate []byte) error {
+	if key == nil {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+	var (
+		res pgconn.CommandTag
+		err error
+	)
+	if valuePredicate == nil {
+		// use insert to make sure there was no previous value before
+		res, err = s.Pool.Exec(ctx, `INSERT INTO `+s.Params.TableName+`(key,value) VALUES($1,$2) ON CONFLICT DO NOTHING`, key, value)
+	} else {
+		// update just in case the previous value was same as predicate value
+		res, err = s.Pool.Exec(ctx, `UPDATE `+s.Params.TableName+` SET value=$2 WHERE key=$1 AND value=$3`, key, value, valuePredicate)
+	}
+	if err != nil {
+		return fmt.Errorf("%w: %s", kv.ErrOperationFailed, err)
+	}
+	if res.RowsAffected() != 1 {
+		return kv.ErrNotFound
+	}
+	return nil
+}
+
+func (s *Store) Delete(ctx context.Context, key []byte) error {
+	if key == nil {
+		return kv.ErrMissingKey
+	}
+	res, err := s.Pool.Exec(ctx, `DELETE FROM `+s.Params.TableName+` WHERE key=$1`, key)
+	if err != nil {
+		return fmt.Errorf("%w: %s", kv.ErrOperationFailed, err)
+	}
+	if res.RowsAffected() != 1 {
+		return kv.ErrNotFound
+	}
+	return nil
+}
+
+func (s *Store) Scan(ctx context.Context, start []byte) (kv.Entries, error) {
+	var (
+		rows pgx.Rows
+		err  error
+	)
+	if start == nil {
+		rows, err = s.Pool.Query(ctx, `SELECT key,value FROM `+s.Params.TableName+` ORDER BY key`)
+	} else {
+		rows, err = s.Pool.Query(ctx, `SELECT key,value FROM `+s.Params.TableName+` WHERE key >= $1 ORDER BY key`, start)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", kv.ErrOperationFailed, err)
+	}
+	return &Entries{
+		rows: rows,
+	}, nil
+}
+
+func (s *Store) Close() {
+	if s.closed {
+		return
+	}
+	s.Pool.Close()
+	s.closed = true
+}
+
+// Next read the next key/value on any error entry will be nil, err will be set by trying to scan the results
+func (e *Entries) Next() bool {
+	e.entry = nil
+	e.err = nil
+	if !e.rows.Next() {
+		return false
+	}
+	var ent kv.Entry
+	if err := e.rows.Scan(&ent.Key, &ent.Value); err != nil {
+		e.err = fmt.Errorf("%w: %s", kv.ErrOperationFailed, err)
+		return false
+	}
+	e.entry = &ent
+	return true
+}
+
+func (e *Entries) Entry() *kv.Entry {
+	return e.entry
+}
+
+// Err return the last scan error or the cursor error
+func (e *Entries) Err() error {
+	if e.err != nil {
+		return e.err
+	}
+	if err := e.rows.Err(); err != nil {
```

Comment:
Why show initiative and check for this error? It will come up in the next read anyway..

Reply:
Actually I don't think so.  The way to iterate in Go is to call `Next` until it returns `false`, then examine `Err` to determine whether `Next` stopped because of end-of-iteration or because of an error.  In this case, `e.rows.Next()` on l.203 will return false, so `Next()` will return false.  But then when the caller checks `Err()` it needs to know (immediately, on this call) that something went wrong.  So the check here is required by Go semantics.